leakcanary-hprof/src/main/java/leakcanary/hprof/HProfParser.java

```diff
@@ -0,0 +1,1109 @@
+            if (type != Type.CHAR) {
+              log(4, String.format("%s[%d]", type, numElements),
+                  elements.size() > 30 ? elements.subList(0, 30) : elements);
+            } else {
+              log(4, String.format("%s[%d]", type, numElements),
+                  "\"" +
+                      elements.stream().map(Object::toString).reduce((acc, e) -> acc + e).get() +
+                      "\""
+              );
+            }
+          }
+
+          primitiveArraySet.add(arrayObjectId);
+
+          break;
+        }
+        case HEAP_DUMP_INFO: {
+          HEAP_DUMP_INFO_COUNT++;
+
+          long objectId = readId(hprofSource);
+          long heapNameStringId = readId(hprofSource);
+          numBytes -= 2 * idSize;
+
+          log(3, "tag", "HEAP DUMP INFO");
+          log(4, "object ID", objectId);
+          log(4, "heap name", getStringName(heapNameStringId));
+
+          break;
+        }
+        case ROOT_INTERNED_STRING: {
+          ROOT_INTERNED_STRING_COUNT++;
+
+          long objectId = readId(hprofSource);
+          numBytes -= idSize;
+
+          log(3, "tag", "ROOT INTERNED STRING");
+          log(4, "object ID", stringMap.get(objectId));
+
+          break;
+        }
+        case ROOT_FINALIZING: {
+          ROOT_FINALIZING_COUNT++;
+
+          long objectId = readId(hprofSource);
+          numBytes -= idSize;
+
+          log(3, "tag", "ROOT FINALIZING");
+          log(4, "object ID", objectId);
+
+          break;
+        }
+        case ROOT_DEBUGGER: {
+          ROOT_DEBUGGER_COUNT++;
+
+          long objectId = readId(hprofSource);
+          numBytes -= idSize;
+
+          if (LOG_ROOTS) {
+            log(3, "tag", "ROOT DEBUGGER");
+            // log(4, "object ID", objectId);
+          }
+
+          break;
+        }
+        default:
+          log(0, "numBytes", numBytes);
+          throw new IllegalArgumentException("Unexpected heap dump record type: " + tag);
+      }
+    }
+  }
+
+  private String getStringName(long stringId) {
+    String stringName = stringMap.get(stringId);
+    return stringName == null ? "???" : stringName;
+  }
+
+  private String getClassName(long classObjectId) {
+    Long classStringId = classMap.get(classObjectId);
+    if (classStringId == null) return "???";
+    return getStringName(classStringId);
+  }
+
+  private String getClassName(int classSerialId) {
+    Long classStringId = classSerialMap.get(classSerialId);
+    if (classStringId == null) return "???";
+    return getStringName(classStringId);
+  }
+
+  private int getTypeSize(Type type) {
+    return type == Type.OBJECT ? idSize : type.numBytes;
+  }
+
+  private Object readTypeElement(BufferedSource hprofSource, Type type) throws IOException {
+    switch (type) {
+      case OBJECT:
+        return readId(hprofSource);
+      case BOOLEAN:
+        return hprofSource.readByte() == 0;
+      case CHAR:
+        return hprofSource.readUtf8(2).charAt(0);
```

Comment:
Is this UTF-16 data? Proably want `readString(StandardCharsets.UTF_16).charAt(0)` then

Reply:
hmm, i forget.  added as a comment for now.